### PR TITLE
Implement position translation with clipping

### PR DESCRIPTION
**Pull request by nathansobo**
This PR demonstrates my preferred approach for dealing with clipping when translating positions. It also restructures some other things since it's early in the project and I didn't want to bother breaking things up into tiny pieces. One big change is that it replaces `consume` and `emit*` with a single function, `transform`. Its parameters are as follows:
- A count of characters to consume from the input
- Optional content for the output (pass through is assumed if this content is null)
- An optional extent for the output, enabling character collapse and newline insertion to be expressed.
- An optional clipping constant. For now we only support open clipping intervals but we'll need to add closed-open to disallow markers at the beginning of soft-wrap indents.

Clipping is implemented in `TransformLayerIterator::seek`, which takes an optional second `clip` argument to indicate whether we should clip forward or backward to the nearest valid position when the seek lands inside a clipped region.

@as-cii Definitely not intending to step on your PR with this one, but I needed to code freely. :seedling:  Would you be interested in finishing this branch out with tests for layers besides hard-tab? This branch probably conflicts textually with your PR but the concepts are pretty portable.

### CODE REVIEW

File: src/transform-layer.coffee [hunk 1]
```diff
@@ -32,3 +38,81 @@ class TransformLayer extends Layer
...
+    @transformBuffer.reset(@position, @sourcePosition)
+    return if position.isZero()
+
+    until @sourcePosition.compare(position) >= 0
+      lastPosition = @position
+      lastSourcePosition = @sourcePosition
+      {done} = @next()
+      break if done
+
+    return if @clipping?
```
Comment (as-cii) [inline src/transform-layer.coffee:107]: Also, please note that clients are calling `seekToSourcePosition` with a clipping parameter which is not being used here.

File: src/transform-layer.coffee [hunk 2]
```diff
@@ -32,3 +38,81 @@ class TransformLayer extends Layer
...
+
+    overshoot = position.column - lastSourcePosition.column
```
Comment (as-cii) [inline src/transform-layer.coffee]: Related to why 0acfae82ae83cdb727dfe9235e658f9d0f1a6fc3 is failing, I think that the overshoot calculation is the culprit. I feel like we should take into account both `x` and `y` when calculating it and to clip to the next row where appropriate.

How do you think we could proceed here?

Comment (nathansobo) [inline src/transform-layer.coffee]: Yeah, I think we should use the methods on `Point` to calculate this. Go ahead and make the change.

File: src/transform-layer.coffee (hunk 1 above)
Comment (nathansobo) [inline src/transform-layer.coffee:107]: :+1:  Go ahead and remove the parameter until we have occasion to use this.

**Comment by nathansobo** [comment]
Thanks for picking this up and pushing it forward. Your changes are looking :sparkles:. I may have time later today to look into the spec failures, but the overshoot calculation is definitely a culprit so that's a good place to start investigating.

**Comment by as-cii** [comment]
> Thanks for picking this up and pushing it forward. Your changes are looking . I may have time later today to look into the spec failures, but the overshoot calculation is definitely a culprit so that's a good place to start investigating.

Working on this stuff is just exciting and I hope to get more and more confident with the code as we progress! :blush: 

I used `Point` to calculate the overshoot, but the problem was that we were calculating it even when `sourcePosition` was equal to the desired position. By the way, we should be good to go now! :hand:

There are still other specs which I copied over from `TextBuffer` but excluded intentionally, because clipping invalid positions hasn't been implemented yet. This last "feature" may also be implemented in terms of `TransformIterator#seek`: since I am not sure this is what we actually want, I sketched out a first implementation (which passes all the tests) in another PR which I am going to open soon. Code-wise it's still a bit messy, but it demonstrates the point I believe. I'll mention you there so that we can have a brief discussion (or just close it if I wrote crappy code :sweat_smile:)

**Pull request merged**